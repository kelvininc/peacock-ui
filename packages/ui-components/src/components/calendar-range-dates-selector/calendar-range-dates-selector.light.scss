@import './calendar-range-dates-selector.night'; // TODO: Check with design team color palette for light theme
