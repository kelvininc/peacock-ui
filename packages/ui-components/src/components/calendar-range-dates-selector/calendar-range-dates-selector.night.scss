@import '../../assets/styles/globals';

kv-calendar::part(day-container) {
	--calendar-day-text-color-in-range: #{kv-color('neutral-0')};
	--calendar-day-background-color-in-range: #{kv-color('neutral-6')};
}
