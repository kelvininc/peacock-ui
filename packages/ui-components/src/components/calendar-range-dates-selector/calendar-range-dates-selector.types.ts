import { EventEmitter } from '@stencil/core';
import { ICalendar } from '../calendar/calendar.types';

export interface ICalendarRangeDatesSelector extends Omit<ICalendar, 'selectedDates'> {
	/** (optional) Currently selected range dates */
	selectedRangeDates?: [] | [string] | [string, string];
}

export interface ICalendarRangeDatesSelectorEvents {
	/** Emitted when range dates are selected */
	selectRangeDates: EventEmitter<ISelectRangeDates>;
}

export interface ISelectRangeDates {
	event: MouseEvent;
	payload: [] | [string] | [string, string];
}
