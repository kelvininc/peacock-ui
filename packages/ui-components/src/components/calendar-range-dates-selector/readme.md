# kv-calendar-range-dates-selector

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-calendar-range-dates-selector></kv-calendar-range-dates-selector>

<!-- With selected ranges date -->
<kv-calendar-range-dates-selector [selectedRangeDates]="['2022-08-04', '2022-09-01']"></kv-calendar-range-dates-selector>

<!-- With initial date -->
<kv-calendar-range-dates-selector initialDate="2022-08-04"></kv-calendar-range-dates-selector>

<!-- With disabled dates-->
<kv-calendar-range-dates-selector [disabledDates]="['2021-12-01', '2021-12-04', '2021-12-05']"></kv-calendar-range-dates-selector>

<!-- With min and max date -->
<kv-calendar-range-dates-selector minDate="2021-12-04" maxDate="2023-02-12"></kv-calendar-range-dates-selector>
```


### React

```tsx
import React from 'react';

import { KvCalendarRangeDatesSelector } from '@kelvininc/react-ui-components';

export const TextFieldExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvCalendarRangeDatesSelector />

		{/*-- With selected date --*/}
		<KvCalendarRangeDatesSelector selectedDateRange={['2022-08-04', '2022-09-01']} />

		{/*-- With initial date --*/}
		<KvCalendarRangeDatesSelector initialDate="2021-12-03" />

		{/*-- With disabled dates --*/}
		<KvCalendarRangeDatesSelector disabledDates={['2021-12-01', '2021-12-04', '2021-12-05']} />

		{/*-- With min and max --*/}
		<KvCalendarRangeDatesSelector minDate="2021-12-04" maxDate="2023-02-12" />
	</>
);
```



## Properties

| Property             | Attribute      | Description                               | Type                                 | Default     |
| -------------------- | -------------- | ----------------------------------------- | ------------------------------------ | ----------- |
| `disabledDates`      | --             | (options) The disabled dates array        | `string[]`                           | `[]`        |
| `initialDate`        | `initial-date` | (optional) Initial date                   | `string`                             | `undefined` |
| `maxDate`            | `max-date`     | (optional) Maximum accepted date          | `string`                             | `undefined` |
| `minDate`            | `min-date`     | (optional) Minimum accepted date          | `string`                             | `undefined` |
| `selectedRangeDates` | --             | (optional) Currently selected range dates | `[] \| [string, string] \| [string]` | `[]`        |


## Events

| Event              | Description                           | Type                             |
| ------------------ | ------------------------------------- | -------------------------------- |
| `selectRangeDates` | Emitted when range dates are selected | `CustomEvent<ISelectRangeDates>` |


## Dependencies

### Used by

 - [kv-range-dates-select-dropdown](../range-dates-select-dropdown)

### Depends on

- [kv-calendar](../calendar)

### Graph
```mermaid
graph TD;
  kv-calendar-range-dates-selector --> kv-calendar
  kv-calendar --> kv-action-button-icon
  kv-calendar --> kv-calendar-day
  kv-action-button-icon --> kv-action-button
  kv-action-button-icon --> kv-icon
  kv-action-button-icon --> kv-badge
  kv-range-dates-select-dropdown --> kv-calendar-range-dates-selector
  style kv-calendar-range-dates-selector fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


