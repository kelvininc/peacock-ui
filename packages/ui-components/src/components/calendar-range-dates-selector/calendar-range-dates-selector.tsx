import { Component, Event, EventEmitter, h, Host, Prop, Watch } from '@stencil/core';
import { formatDate, fromISOToMoment, getDatesBetweenRange, isDateBefore, isDateSame, isDateValid } from '../../utils/date.helper';
import { IClickDateEvent, SelectedRange } from '../calendar/calendar.types';
import { ICalendarRangeDatesSelector, ICalendarRangeDatesSelectorEvents, ISelectRangeDates } from './calendar-range-dates-selector.types';

@Component({
	tag: 'kv-calendar-range-dates-selector',
	styleUrls: {
		night: 'calendar-range-dates-selector.night.scss',
		light: 'calendar-range-dates-selector.light.scss'
	},
	shadow: true
})
export class KvCalendarRangeDatesSelector implements ICalendarRangeDatesSelector, ICalendarRangeDatesSelectorEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedRangeDates?: SelectedRange = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) minDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) maxDate?: string;

	/** @inheritdoc */
	@Event() selectRangeDates: EventEmitter<ISelectRangeDates>;

	@Watch('selectedRangeDates')
	validateRangeDates(newRangeDates: SelectedRange | undefined) {
		if (newRangeDates !== undefined && newRangeDates.length > 0) {
			const [selectedStartDate, selectedEndDate] = newRangeDates;

			if (selectedStartDate !== undefined && !isDateValid(selectedStartDate)) {
				throw new Error('Range selected start date should be a valid date');
			}

			if (selectedEndDate !== undefined && !isDateValid(selectedEndDate)) {
				throw new Error('Range selected end date should be a valid date');
			}
		}
	}

	componentWillLoad() {
		this.validateRangeDates(this.selectedRangeDates);
	}

	private onClickDate = ({ detail: { event, payload: date } }: CustomEvent<IClickDateEvent>): void => {
		const clickedDateMoment = fromISOToMoment(date);

		const [selectedStartDate, selectedEndDate] = this.getSelectedRangeDates();

		// check if start date is not selected
		if (selectedStartDate === undefined) {
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// check if clicked date is the same as the selected date
		if (isDateSame(clickedDateMoment, selectedStartDate)) {
			// check if end date is not selected
			if (selectedEndDate === undefined) {
				this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment), formatDate(clickedDateMoment)] });
				return;
			}

			this.selectRangeDates.emit({ event, payload: [] });
			return;
		}

		// check if end date is selected
		if (selectedEndDate !== undefined) {
			// reset end date and set
			// start date to the clicked date
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// check if clicked day is before the start date
		if (isDateBefore(clickedDateMoment, selectedStartDate)) {
			// reset end date and set
			// start date to clicked date
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// the clicked date is after the clicked date
		// set end date to the clicked day
		this.selectRangeDates.emit({ event, payload: [formatDate(selectedStartDate), formatDate(clickedDateMoment)] });
		return;
	};

	public getSelectedDates = (): string[] | undefined => {
		const [selectedStartDate, selectedEndDate] = this.getSelectedRangeDates();

		if (selectedStartDate === undefined) {
			return;
		}

		if (selectedEndDate === undefined) {
			return [selectedStartDate];
		}

		return getDatesBetweenRange(selectedStartDate, selectedEndDate);
	};

	private getSelectedRangeDates = () => this.selectedRangeDates ?? [];

	render() {
		const [selectedStartDate] = this.getSelectedRangeDates();
		const selectedDates = this.getSelectedDates();

		return (
			<Host>
				<kv-calendar
					initialDate={this.initialDate ?? selectedStartDate}
					onClickDate={this.onClickDate}
					selectedDates={selectedDates}
					disabledDates={this.disabledDates}
					minDate={this.minDate}
					maxDate={this.maxDate}
					exportparts="calendar-container,month-container,day-container"
				/>
			</Host>
		);
	}
}
