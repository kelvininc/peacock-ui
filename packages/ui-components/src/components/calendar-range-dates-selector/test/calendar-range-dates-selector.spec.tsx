import { h, SpecPage } from '@stencil/core/internal';
import { KvCalendarRangeDatesSelector } from '../calendar-range-dates-selector';
import { newSpecPage } from '@stencil/core/testing';

describe('Calendar Range Dates Selector (unit tests)', () => {
	let page: SpecPage;
	let component: KvCalendarRangeDatesSelector;

	describe('when selected dates is passed', () => {
		describe('and start and end dates are valid', () => {
			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarRangeDatesSelector],
					template: () => <kv-calendar-range-dates-selector selectedRangeDates={['2020-04-02', '2020-04-05']}></kv-calendar-range-dates-selector>
				});
				component = page.rootInstance;
			});

			it('should match the snapshot', () => {
				expect(page.root).toMatchSnapshot();
			});
		});

		describe('and start date is invalid', () => {
			it('should throw a validation error', async () => {
				await expect(
					newSpecPage({
						components: [KvCalendarRangeDatesSelector],
						template: () => <kv-calendar-range-dates-selector selectedRangeDates={['invalid-date', '2020-04-05']}></kv-calendar-range-dates-selector>
					})
				).rejects.toThrow('Range selected start date should be a valid date');
			});
		});

		describe('and end date is invalid', () => {
			it('should throw a validation error', async () => {
				await expect(
					newSpecPage({
						components: [KvCalendarRangeDatesSelector],
						template: () => <kv-calendar-range-dates-selector selectedRangeDates={['2020-04-02', 'invalid-date']}></kv-calendar-range-dates-selector>
					})
				).rejects.toThrow('Range selected end date should be a valid date');
			});
		});
	});

	describe('when `getSelectedDates` is called', () => {
		describe('and there is no selected range dates', () => {
			let result: string[] | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarRangeDatesSelector],
					html: '<kv-calendar-range-dates-selector></kv-calendar-range-dates-selector>'
				});
				component = page.rootInstance;
				result = component.getSelectedDates();
			});

			it('should return undefined', () => {
				expect(result).toBeUndefined();
			});
		});

		describe('and there is no selected end range date', () => {
			let result: string[] | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarRangeDatesSelector],
					template: () => <kv-calendar-range-dates-selector selectedRangeDates={['2020-04-02']}></kv-calendar-range-dates-selector>
				});
				component = page.rootInstance;
				result = component.getSelectedDates();
			});

			it('should return an array with the selected start date', () => {
				expect(result).toEqual(['2020-04-02']);
			});
		});

		describe('and there is a selected range dates', () => {
			let result: string[] | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvCalendarRangeDatesSelector],
					template: () => <kv-calendar-range-dates-selector selectedRangeDates={['2020-04-02', '2020-04-05']}></kv-calendar-range-dates-selector>
				});
				component = page.rootInstance;
				result = component.getSelectedDates();
			});

			it('should return an array with all dates between the range', () => {
				const expectedResult = ['2020-04-02', '2020-04-03', '2020-04-04', '2020-04-05'];
				expect(result).toEqual(expectedResult);
			});
		});
	});
});
