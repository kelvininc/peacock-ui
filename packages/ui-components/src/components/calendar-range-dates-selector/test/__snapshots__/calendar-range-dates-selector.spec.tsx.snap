// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Range Dates Selector (unit tests) when selected dates is passed and start and end dates are valid should match the snapshot 1`] = `
<kv-calendar-range-dates-selector>
  <mock:shadow-root>
    <kv-calendar exportparts="calendar-container,month-container,day-container" initialdate="2020-04-02"></kv-calendar>
  </mock:shadow-root>
</kv-calendar-range-dates-selector>
`;
